### PR TITLE
World snapshot and multiple compressed snapshots per module

### PR DESCRIPTION
**Pull request by miloszm**
New snapshot related functionality:
1. multiple snapshots per module, i.e., a collection of snapshots per module instance
2. world's scoped snapshot id and world snapshot support
3. collections of module snapshots for world snapshot and snapshots of a module
4. absolute and relative snapshots support (a relative snapshot is synonymous to a diff'ed and compressed snapshot)

This PR fixes issue 33

### CODE REVIEW

File: hatchery/src/lib.rs [hunk 1]
```diff
@@ -9,11 +9,14 @@ mod error;
...
 mod memory;
 mod snapshot;
+mod snapshot_bag;
 mod storage_helpers;
 mod world;
+mod world_snapshot;
 
 pub use error::Error;
 pub use world::{Event, Receipt, World};
+pub use world_snapshot::WorldSnapshotId;
```
Comment (ureeves) [inline hatchery/src/lib.rs:19]: I suggest switching it around, and do the following renames:
```text
`SnapshotId` -> `ModuleSnapshotId`
`WorldSnapshotId` -> `SnapshotId`
```

File: hatchery/src/world_snapshot.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    Archive,
+    Serialize,
+    Deserialize,
+    PartialOrd,
+    Ord,
+    Hash,
+    Clone,
+    Copy,
+)]
+pub struct WorldSnapshotId([u8; WORLD_SNAPSHOT_ID_BYTES]);
```
Comment (ureeves) [inline hatchery/src/world_snapshot.rs:30]: This should have documentation since it is exported

**Review by ureeves — CHANGES_REQUESTED** [review]
Looks great! I suggest some changes to the exported API.

File: hatchery/src/lib.rs (hunk 1 above)
Comment (miloszm) [reply hatchery/src/lib.rs:19]: good suggestion, will do that

File: hatchery/src/world_snapshot.rs (hunk 2 above)
Comment (miloszm) [reply hatchery/src/world_snapshot.rs:30]: ok

File: hatchery/src/snapshot_bag.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            // accu and snapshot are both uncompressed
+            // compressing snapshot against the top
+            snapshot.capture_diff(&top_snapshot, memory_path)?;
+            // snapshot is compressed but accu keeps the uncompressed copy
+            // top is an uncompressed version of most recent snapshot
+            top_snapshot.capture(&accu_snapshot)?;
+            Ok(self.ids.len() - 1)
+        }
+    }
+    pub fn restore_snapshot(
```
Comment (krl) [inline hatchery/src/snapshot_bag.rs:53]: empty line between method blocks

File: hatchery/src/world.rs [hunk 4]
```diff
@@ -82,41 +85,60 @@ impl World {
...
+            w.snapshots.get(world_snapshot_id).ok_or_else(|| {
+                SnapshotError(String::from("world snapshot id not found"))
+            })?;
+        world_snapshot.restore_snapshots(self)?;
         Ok(())
     }
     pub fn memory_path(&self, module_id: &ModuleId) -> PathBuf {
         self.storage_path().join(module_id_to_name(*module_id))
     }
+    pub fn restore_snapshot_with_index(
```
Comment (krl) [inline hatchery/src/world.rs:124]: The consumers should not have to know about snapshot indexes. (using snapshot_index in a pub method)

File: hatchery/src/snapshot_bag.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        &self,
+        snapshot_index: usize,
```
Comment (krl) [inline hatchery/src/snapshot_bag.rs:55]: The consumers should not have to know about snapshot indexes. (using snapshot_index in a pub method)

File: hatchery/src/storage_helpers.rs [hunk 6]
```diff
@@ -18,7 +18,7 @@ pub fn module_id_to_name(module_id: ModuleId) -> String {
     format!("{}", ByteArrayWrapper(module_id.as_bytes()))
 }
 
-pub fn snapshot_id_to_name(snapshot_id: SnapshotId) -> String {
+pub fn snapshot_id_to_name(snapshot_id: ModuleSnapshotId) -> String {
```
Comment (krl) [inline hatchery/src/storage_helpers.rs:21]: This should not be public.

File: hatchery/src/world_snapshot.rs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+#[derive(Debug)]
+pub struct WorldSnapshot {
+    id: SnapshotId,
+    snapshot_indices: BTreeMap<ModuleId, usize>,
+}
+
+impl WorldSnapshot {
+    pub fn new() -> Self {
```
Comment (krl) [inline hatchery/src/world_snapshot.rs:59]: this should not be public.

**Review by krl — CHANGES_REQUESTED** [review]

File: hatchery/src/instance.rs [hunk 8]
```diff
@@ -224,11 +224,11 @@ impl Instance {
         self.id
     }
 
-    pub(crate) fn set_snapshot_id(&mut self, snapshot_id: SnapshotId) {
-        self.snapshot_id = Some(snapshot_id);
+    pub(crate) fn module_snapshot_bag(&self) -> &ModuleSnapshotBag {
+        &self.module_snapshot_bag
     }
```
Comment (krl) [inline hatchery/src/instance.rs:229]: please use blank lines between method bodies

File: hatchery/Cargo.toml [hunk 9]
```diff
@@ -17,3 +17,7 @@ dallo = { path = "../dallo" }
 blake3 = "1.3.1"
 parking_lot = "0.12.1"
 tempfile = "3.2.0"
+qbsdiff = "1.4"
+zstd = "0.9"
+byteorder = "1.4.3"
```
Comment (krl) [inline hatchery/Cargo.toml:22]: Use rkyv instead of byteorder for serializing.

File: hatchery/src/snapshot/module_snapshot.rs [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+pub struct ModuleSnapshot {
+    path: PathBuf,
+    id: ModuleSnapshotId,
+}
+
+impl ModuleSnapshot {
+    pub(crate) fn new(memory_path: &MemoryPath) -> Result<Self, Error> {
+        let module_snapshot_id: ModuleSnapshotId = ModuleSnapshotId::from(
+            *blake3::hash(memory_path.read()?.as_slice()).as_bytes(),
```
Comment (krl) [inline hatchery/src/snapshot/module_snapshot.rs:103]: This hashing includes the argument buffer, which should be ignored when hashing.

**Review by krl — COMMENTED** [review]
I'd like to see more comprehensive testing.

Also, the argbuf should not be included in the hash of the module, please also include a test for this.

**Comment by miloszm** [comment]
closing this PR as it was superseded by PR 88, the branch contains diff-ing functionality to be used in the future, so the branch should be retained for future use

**Pull request closed without merging**